README.md
#sqlite3_driver
==========

SQLite database engine driver.

Version: 0.0.4

The goal of the Dash effort is ultimately to replace JavaScript as the lingua franca of web development on the open web platform.

**Notice:**

This driver uses [sqlite3_bindings](https://pub.dartlang.org/packages/sqlite3_bindings) software.

Please read carefully information about usage of [sqlite3_bindings](https://pub.dartlang.org/packages/sqlite3_bindings).

**Example of usage:**

```dart
import 'dart:async';

import 'package:sqlite3_driver/sqlite3_driver.dart';

Future main() async {
  var sw = new Stopwatch();
  // ****************
  // Create connection
  // ****************
  _start(sw, "Create connection");
  var connection = new Sqlite3Connection(":memory:");
  _stop(sw, "Create connection");

  // ****************
  // Open connection
  // ****************
  _start(sw, "Open connection");
  await connection.open();
  _stop(sw, "Open connection");

  try {
    // ****************
    // Create table
    // ****************
    _start(sw, "Create table");
    var stmtCreate = new Sqlite3Statement(connection, sqlCreateTable);
    await stmtCreate.execute();
    _stop(sw, "Create table");

    // ****************
    // Insert data
    // ****************
    var count = 10000;
    _start(sw, "Insert $count rows");
    var stmtInsert = new Sqlite3Statement(connection, sqlInsertData);
    for (var i = 0; i < count; i++) {
      var parameters = {};
      parameters[":ID"] = i;
      parameters[":NAME"] = "Name";
      parameters[":AGE"] = 25;
      parameters[":ADDRESS"] = "Address";
      parameters[":SALARY"] = 10000.0;
      parameters[":DATA"] = [0, i];
      await stmtInsert.execute(parameters);
    }

    _stop(sw, "Insert $count rows");

    // ****************
    // Read data
    // ****************
    _start(sw, "Read $count rows");
    var stmtSelect = new Sqlite3Statement(connection, sqlReadData);
    var stream = await stmtSelect.executeQuery({":AGE": 25});
    await for (var row in stream) {
      var id = row["ID"];
    }

    _stop(sw, "Read $count rows");

    // ****************
    // Delete data
    // ****************
    _start(sw, "Delete $count rows");
    var stmtDelete = new Sqlite3Statement(connection, sqlDeleteData);
    await stmtDelete.execute({":AGE": 25});
    var affectedRows = await connection.affectedRows;
    _stop(sw, "Delete $affectedRows rows");
  } finally {
    connection.close();
  }
}

void _start(Stopwatch sw, String message) {
  print("$message...");
  sw.reset();
  sw.start();
}

void _stop(Stopwatch sw, String message) {
  sw.stop();
  print("$message: ${sw.elapsedMilliseconds / 1000} sec");
}

String sqlCreateTable = """
CREATE TABLE COMPANY (
ID INT PRIMARY KEY NOT NULL,
NAME TEXT NOT NULL,
AGE INT NOT NULL,
ADDRESS CHAR(50),
SALARY REAL,
DATA BLOB);
""";

String sqlDeleteData = """
DELETE
FROM COMPANY
WHERE AGE = :AGE;
""";

String sqlInsertData = """
INSERT INTO COMPANY (ID, NAME, AGE, ADDRESS, SALARY, DATA)
VALUES (:ID, :NAME, :AGE, :ADDRESS, :SALARY, :DATA);
""";

String sqlReadData = """
SELECT *
FROM COMPANY
WHERE AGE = :AGE;
""";

```


CHANGELOG.md
## 0.0.4

- Adapting to the new version of `sqlite3_bindings`

## 0.0.3

- Updated example of usage

## 0.0.2

- Added example of usage

## 0.0.1

- Initial release



test/test.dart
import 'dart:async';import 'dart:io';import 'package:sqlite3_driver/sqlite3_driver.dart';import 'package:test/test.dart';Future main() async {  test("SQLite database engine driver", () async {    var filename = ":memory:";    var file = new File(filename);    if (file.existsSync()) {      file.deleteSync();    }    var conn = new Sqlite3Connection(filename);    try {      await conn.open();      var stmtCreateTable = new Sqlite3Statement(conn, sqlCreateTable);      await stmtCreateTable.execute();      var stmtInsertData = new Sqlite3Statement(conn, sqlInsertData);      for (var data in intilaData) {        await stmtInsertData.execute(data);      }      var stmtSelectData = new Sqlite3Statement(conn, sqlSelectData);      var rows = await stmtSelectData.executeQuery();      var index = 0;      await for (var row in rows) {        var original = intilaData[index];        for (var i = 0; i < row.length; i++) {          expect(row[i], original[i]);        }        var i = 0;        for (var name in row.names) {          expect(row[name], original[i++]);        }        index++;      }      var stmtUpdateData = new Sqlite3Statement(conn, sqlUpdateData);      await stmtUpdateData.executeQuery({":AGE": 25, ":NEW_AGE": 26});      var changes = await conn.affectedRows;      expect(changes, 2);      var stmtSelectCount = new Sqlite3Statement(conn, sqlSelectCount);      rows = await stmtSelectCount.executeQuery([26]);      var count = 0;      await for (var row in rows) {        count += row[0];      }      expect(count, 2);    } finally {      await conn.close();    }  });}List<List> intilaData = [  [1, 'Gérard', 32, 'California', 20000.00, []],  [2, 'Allen', 25, 'Texas', 15000.00, null],  [3, 'Teddy', 23, 'Norway', 20000.00, [1, 2, 3]],  [4, 'Mark', 25, 'Rich-Mond ', 65000.00, [123]],];String sqlCreateTable = """CREATE TABLE COMPANY (ID INT PRIMARY KEY NOT NULL,NAME TEXT NOT NULL,AGE INT NOT NULL,ADDRESS CHAR(50),SALARY REAL,DATA BLOB);""";String sqlInsertData = """INSERT INTO COMPANY (ID, NAME, AGE, ADDRESS, SALARY, DATA)VALUES (:ID, :NAME, :AGE, :ADDRESS, :SALARY, :DATA);""";String sqlSelectCount = """SELECT COUNT(*)FROM COMPANYWHERE AGE = ?""";String sqlSelectData = """SELECT * FROM COMPANY""";String sqlUpdateData = """UPDATE COMPANYSET AGE = :NEW_AGEWHERE AGE = :AGE""";

lib/src/sqlite3_driver/sqlite3_driver.dart
part of sqlite3_driver;

class Sqlite3Connection implements SqlConnection {
  Set<SqlStatement> _statements = new Set<SqlStatement>();

  String _filename;

  _SqliteHelper _helper;

  BinaryData _pDb;

  Sqlite3Connection(String filename) {
    if (filename == null) {
      new ArgumentError.notNull("filename");
    }

    _filename = filename;
    _helper = new _SqliteHelper();
  }

  Future<int> get affectedRows async {
    if (_pDb == null) {
      return 0;
    }

    return _helper.library.sqlite3_changes(_pDb);
  }

  bool get isOpen {
    return _pDb != null;
  }

  Future<int> get lastInsertId async {
    if (_pDb == null) {
      return 0;
    }

    return _helper.library.sqlite3_last_insert_rowid(_pDb);
  }

  Future close() async {
    if (_pDb == null) {
      return;
    }

    for (var statement in _statements.toList()) {
      await statement.free();
    }

    var errorCode = _helper.library.sqlite3_close(_pDb);
    _helper.checkError(errorCode, this);
    _pDb = null;
  }

  Future open() async {
    if (_pDb != null) {
      return;
    }

    var ppDb = _helper.types["sqlite3*"].alloc(null);
    var pFilename = _helper.allocString16(_filename);
    var errorCode = _helper.library.sqlite3_open16(pFilename, ppDb);
    _pDb = ppDb.value;
    if (_pDb.isNullPtr) {
      throw new SqlException("Not enough memory");
    }

    _helper.checkError(errorCode, this);
  }
}

class Sqlite3Statement implements SqlStatement {
  final String commandText;

  final Sqlite3Connection connection;

  _SqliteHelper _helper;

  BinaryData _pStmt;

  Sqlite3Statement(this.connection, this.commandText) {
    if (connection == null) {
      throw new ArgumentError.notNull("connection");
    }

    if (commandText == null) {
      throw new ArgumentError.notNull("commandText");
    }

    if (commandText.trim().isEmpty) {
      throw new ArgumentError("Command text is empty");
    }

    if (!connection.isOpen) {
      throw new SqlException("Connection is not open");
    }

    _helper = new _SqliteHelper();
    var ppStmt = _helper.types["sqlite3_stmt*"].alloc(null);
    var pVoid = _helper.types["void*"].nullPtr;
    var pCommandText = _helper.allocString16(commandText);
    var errorCode = _helper.library
        .sqlite3_prepare16_v2(connection._pDb, pCommandText, -1, ppStmt, pVoid);
    _helper.checkError(errorCode, connection);
    _pStmt = ppStmt.value;
    connection._statements.add(this);
  }

  Future execute([dynamic parameters]) async {
    _checkConnection();
    _bindParameters(parameters);
    var errorCode = _helper.library.sqlite3_step(_pStmt);
    _checkResult(errorCode);
    _reset();
    return null;
  }

  Future<Stream<SqlDataRow>> executeQuery([dynamic parameters]) async {
    _checkConnection();
    _bindParameters(parameters);
    var controller = new StreamController<SqlDataRow>();
    var columnCount = _helper.library.sqlite3_column_count(_pStmt);
    var columns = new List<String>(columnCount);
    int errorCode;
    var int16_t = _helper.types["int16_t"];
    for (var i = 0; i < columnCount; i++) {
      var name =
          _helper.readString(_helper.library.sqlite3_column_name(_pStmt, i));
      columns[i] = name;
    }

    while ((errorCode = _helper.library.sqlite3_step(_pStmt)) ==
        Sqlite3Def.SQLITE_ROW) {
      var data = <String, dynamic>{};
      for (var i = 0; i < columnCount; i++) {
        var columnType = _helper.library.sqlite3_column_type(_pStmt, i);
        var value;
        switch (columnType) {
          case Sqlite3Def.SQLITE_BLOB:
            var numberOfBytes =
                _helper.library.sqlite3_column_bytes(_pStmt, i);
            var pBuffer = _helper.library.sqlite3_column_blob(_pStmt, i);
            var aChar = _helper.types["char[$numberOfBytes]"]
                .extern(pBuffer.base, pBuffer.offset);
            value = aChar.value;
            break;
          case Sqlite3Def.SQLITE_FLOAT:
            value = _helper.library.sqlite3_column_double(_pStmt, i);
            break;
          case Sqlite3Def.SQLITE_INTEGER:
            value = _helper.library.sqlite3_column_int64(_pStmt, i);
            break;
          case Sqlite3Def.SQLITE3_TEXT:
            var pData = _helper.library.sqlite3_column_text16(_pStmt, i);
            value = _helper.readString16(pData);
            break;
          case Sqlite3Def.SQLITE_NULL:
            value = null;
            break;
          default:
            throw new SqlException(
                "Unsupported data type '$columnType' of column $i");
        }

        data[columns[i]] = value;
      }

      controller.add(new _SqliteDataRow(data));
    }

    _checkResult(errorCode);
    _reset();
    controller.close();
    return controller.stream;
  }

  Future free() async {
    if (_pStmt == null) {
      return;
    }

    _helper.library.sqlite3_finalize(_pStmt);
    connection._statements.remove(this);
  }

  Future prepare() async {
    if (_pStmt != null) {
      return;
    }
  }

  void _bindParameter(int index, dynamic parameter) {
    index++;
    int errorCode;
    if (parameter == null) {
      errorCode = _helper.library.sqlite3_bind_null(_pStmt, index);
    } else if (parameter is int) {
      errorCode = _helper.library.sqlite3_bind_int64(_pStmt, index, parameter);
    } else if (parameter is double) {
      errorCode =
          _helper.library.sqlite3_bind_double(_pStmt, index, parameter);
    } else if (parameter is String) {
      var pData = _helper.allocString16(parameter);
      // TODO: Optimize
      var f = _helper.types["void*"].extern(Sqlite3Def.SQLITE_TRANSIENT);
      errorCode =
          _helper.library.sqlite3_bind_text16(_pStmt, index, pData, -1, f);
    } else if (parameter is List) {
      var length = parameter.length;
      if (length == 0) {
        errorCode = _helper.library.sqlite3_bind_zeroblob(_pStmt, index, -1);
      } else {
        var pData = _helper.types["uint8_t[$length]"].alloc(parameter);
        // TODO: Optimize
        var f = _helper.types["void*"].extern(Sqlite3Def.SQLITE_TRANSIENT);
        errorCode =_helper.library.sqlite3_bind_blob(_pStmt, index, pData, length, f);
      }

    } else {
      throw new ArgumentError(
          "Unsupported type of parameter '$index': ${parameter.runtimeType} ");
    }

    _helper.checkError(errorCode, connection);
  }

  void _bindParameters(dynamic parameters) {
    if (parameters == null) {
      return;
    }

    if (parameters is List) {
      _bindParameterFromList(parameters);
      return;
    }

    if (parameters is Map) {
      _bindParameterFromMap(parameters);
      return;
    }

    throw new ArgumentError.value(parameters, "parameters");
  }

  void _bindParameterFromList(List parameters) {
    var length = parameters.length;
    for (var i = 0; i < length; i++) {
      var parameter = parameters[i];
      _bindParameter(i, parameter);
    }
  }

  void _bindParameterFromMap(Map parameters) {
    for (var name in parameters.keys) {
      // TODO: Convert parameter in UTF-8
      var index = _helper.library.sqlite3_bind_parameter_index(_pStmt, name);
      if (index == 0) {
        throw new ArgumentError("Parameter '$name' not found");
      }

      _bindParameter(index - 1, parameters[name]);
    }
  }

  void _checkConnection() {
    if (!connection.isOpen) {
      throw new SqlException("Connection is not open");
    }

    if (_pStmt == null) {
      throw new SqlException("Statement already finalized");
    }
  }

  void _checkResult(int errorCode) {
    if (errorCode != Sqlite3Def.SQLITE_DONE) {
      _helper.checkError(errorCode, connection);
    }
  }

  void _reset() {
    _helper.library.sqlite3_reset(_pStmt);
  }
}

class _SqliteDataRow implements SqlDataRow {
  final Map<String, dynamic> _data;

  List _list;

  _SqliteDataRow(this._data);

  int get length {
    return _data.length;
  }

  Iterable<String> get names {
    return _data.keys;
  }

  dynamic operator [](dynamic key) {
    if (key == null) {
      throw new ArgumentError.notNull("key");
    }

    if (key is String) {
      if (_data.containsKey(key)) {
        return _data[key];
      } else {
        throw new ArgumentError("Invalid key '$key'");
      }
    } else if (key is int) {
      if (_list == null) {
        _list = _data.values.toList();
      }

      var length = _list.length;
      if (key >= 0 && key < length) {
        return _list[key];
      } else {
        throw new RangeError.range(key, 0, length - 1, "key");
      }
    } else {
      throw new ArgumentError.value(key, "key");
    }
  }
}

class _SqliteHelper {
  static _SqliteHelper _instance;

  Sqlite3Lib _library;

  BinaryTypeHelper _helper;

  BinaryTypes _types;

  factory _SqliteHelper() {
    if (_instance == null) {
      _instance = new _SqliteHelper._internal();
    }

    return _instance;
  }

  _SqliteHelper._internal() {
    _types = new BinaryTypes();
    _helper = new BinaryTypeHelper(types);
    _helper.addHeaders(SQLITE3_HEADERS);
    _helper.addHeaders(LIBC_HEADERS);
    _library = loadSqlite3Library(types);
  }

  Sqlite3Lib get library {
    return _library;
  }

  BinaryTypeHelper get helper {
    return _helper;
  }

  BinaryTypes get types {
    return _types;
  }

  BinaryData allocString16(String string) {
    // TODO: Optimize
    var int16_t = types["int16_t"];
    return helper.allocString(string, type: int16_t);
  }

  void checkError(int errorCode, Sqlite3Connection connection) {
    String message;
    if (errorCode != Sqlite3Def.SQLITE_OK) {
      if (connection._pDb != null) {
        message = readString(_library.sqlite3_errmsg(connection._pDb));
      }

      throw new SqlException(message);
    }
  }

  String readString(BinaryData data) {
    return helper.readString(data);
  }

  String readString16(BinaryData data) {
    // TODO: Optimize
    var int16_t = types["int16_t"];
    var ptr = int16_t.extern(data.base, data.offset);
    return helper.readString(ptr);
  }
}


example/example.dart
import 'dart:async';

import 'package:sqlite3_driver/sqlite3_driver.dart';

Future main() async {
  var sw = new Stopwatch();
  // ****************
  // Create connection
  // ****************
  _start(sw, "Create connection");
  var connection = new Sqlite3Connection(":memory:");
  _stop(sw, "Create connection");

  // ****************
  // Open connection
  // ****************
  _start(sw, "Open connection");
  await connection.open();
  _stop(sw, "Open connection");

  try {
    // ****************
    // Create table
    // ****************
    _start(sw, "Create table");
    var stmtCreate = new Sqlite3Statement(connection, sqlCreateTable);
    await stmtCreate.execute();
    _stop(sw, "Create table");

    // ****************
    // Insert data
    // ****************
    var count = 10000;
    _start(sw, "Insert $count rows");
    var stmtInsert = new Sqlite3Statement(connection, sqlInsertData);
    for (var i = 0; i < count; i++) {
      var parameters = {};
      parameters[":ID"] = i;
      parameters[":NAME"] = "Name";
      parameters[":AGE"] = 25;
      parameters[":ADDRESS"] = "Address";
      parameters[":SALARY"] = 10000.0;
      parameters[":DATA"] = [0, i];
      await stmtInsert.execute(parameters);
    }

    _stop(sw, "Insert $count rows");

    // ****************
    // Read data
    // ****************
    _start(sw, "Read $count rows");
    var stmtSelect = new Sqlite3Statement(connection, sqlReadData);
    var stream = await stmtSelect.executeQuery({":AGE": 25});
    await for (var row in stream) {
      var id = row["ID"];
    }

    _stop(sw, "Read $count rows");

    // ****************
    // Delete data
    // ****************
    _start(sw, "Delete $count rows");
    var stmtDelete = new Sqlite3Statement(connection, sqlDeleteData);
    await stmtDelete.execute({":AGE": 25});
    var affectedRows = await connection.affectedRows;
    _stop(sw, "Delete $affectedRows rows");
  } finally {
    connection.close();
  }
}

void _start(Stopwatch sw, String message) {
  print("$message...");
  sw.reset();
  sw.start();
}

void _stop(Stopwatch sw, String message) {
  sw.stop();
  print("$message: ${sw.elapsedMilliseconds / 1000} sec");
}

String sqlCreateTable = """
CREATE TABLE COMPANY (
ID INT PRIMARY KEY NOT NULL,
NAME TEXT NOT NULL,
AGE INT NOT NULL,
ADDRESS CHAR(50),
SALARY REAL,
DATA BLOB);
""";

String sqlDeleteData = """
DELETE
FROM COMPANY
WHERE AGE = :AGE;
""";

String sqlInsertData = """
INSERT INTO COMPANY (ID, NAME, AGE, ADDRESS, SALARY, DATA)
VALUES (:ID, :NAME, :AGE, :ADDRESS, :SALARY, :DATA);
""";

String sqlReadData = """
SELECT *
FROM COMPANY
WHERE AGE = :AGE;
""";
